IGNITE-13056 Move indexes to core

timoninmaxim opened the pull request:
Moving indexes from indexing to core module.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/AbstractIndex.java:
```diff
@@ -15,25 +15,31 @@
  * limitations under the License.
  */
 
-package org.apache.ignite.internal.processors.query.h2.database.inlinecolumn;
```
Actually I don't know why GitHub creates the link StringIgnoreCaseInlineIndexColumn -> AbstractIndex. `git log` for AbstractIndex is the only commit. I tried to git mv, git rm, create file from scratch again. But nothing helps and GitHub still show this again. 

The same time Intellij IDEA shows history of this file differently, it saw it as GridQueryRowCacheCleaner -> AbstractIndex.

```
% git log modules/core/src/main/java/org/apache/ignite/cache/query/index/AbstractIndex.java
commit cee7855760b3fa1867a9e44e52be4ef786f108ad
Author: Maksim Timonin <timonin.maxim@gmail.com>
Date:   Thu Dec 10 01:25:18 2020 +0300

    Rebuild indexes, fix tests
```

If somebody have any idea how to fix this, please share this.

korlov42 reviewed (changes requested):
Maxim, first of all thank you for your huge effort!

To be honest, I don't like the idea to expose some part of indexing API in public space. And the reasons are follow:
1) current API is not clean enough to be part of public API - somewhere it exposes internal classes as type of return, somewhere requires internal classes as input parameters, and even implements internal interfaces (see SingleCursor)
2) current API is not generic enough -  pieces of BPlusTree index appears out here and there (like segments and data link for instance). 

Some more comments you would find out below, and others will be available on next review (unfortunately, it's quite hard to cover such a big PR for the first attempt)

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/AbstractIndex.java:
```diff
@@ -15,25 +15,31 @@
+ * Abstract class for all Index implementations.
  */
-public class StringIgnoreCaseInlineIndexColumn extends StringInlineIndexColumn {
+public abstract class AbstractIndex implements Index {
```
Currently there are descenders: GeoSpatialIndexImpl, InlineIndexImpl.

korlov42 replied:
It appears a 'geospatial' module was ignored in the project. Ok then

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/Index.java:
```diff
@@ -0,0 +1,83 @@
+     *
+     * @return {@code True} if replaced existing row.
+     */
+    public boolean putx(CacheDataRow row) throws IgniteCheckedException;
```
The previous implementation of H2TreeIndex used the `putx` method only. It did have the `put` method, but in all places it's not used returned value, just checks it for existing, so it's similar to `putx` logic. So it looks like it's just not required. I can add it to make API more wide, but as it won't be a part of public for a while API, do we need it?

korlov42 replied:
well, I always interpret the `-x` suffix as `extended` (for example, with or without additional checks), and to have an extended version of something, you have to have the general one first. But it's just my intuition. And with these in mind I'm OK to leave this as is. Furthermore, since it will be moved to the internal space, we will be able to rename it late if needed.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/IndexName.java:
```diff
@@ -0,0 +1,92 @@
+/**
+ * Represents list of names that fully describes index domain (schema, cache, table, index).
+ */
+public class IndexName implements Serializable {
```
It's not actually correct for case of PR indexes. Index name for any PK will be "_key_PK". So for making purely unique index name we should add table name. 

CacheName may be unnecessary, but I add it for non-sql indexes where schema name is null.

korlov42 replied:
Hm, I forgot about PK. Ok, let's leave it as is for now

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/SingleCursor.java:
```diff
@@ -0,0 +1,52 @@
+
+    /** {@inheritDoc} */
+    @Override public T get() throws IgniteCheckedException {
+        if (currIdx.get() <= 0)
```
Yes, it looks pretty strange. But it's the same logic as org.h2.index.SingleRowCursor does ([Link](https://github.com/h2database/h2database/blob/version-1.4.197/h2/src/main/org/h2/index/SingleRowCursor.java))

Also we have some tests (or logic) that differently works with SingleCursor. Some of them just use get, some of them make `next -> get`

korlov42 replied:
but now it extends a `GridCursor` which claims that `Initial state must be "before first"`. Current implementation violates our own contract

timoninmaxim replied:
I've run tests again, for queries like "select MAX(col) from table" H2 just use get() without next(). So this is a reason to implement it this way. I can add a note about in javadoc about it and create a bug in Jira for futher development of Calcite. Alternative is create a new interface of Cursor. I'd like to stay with GridCursor interface. WDYT?

timoninmaxim replied:
Create a ticket for that. https://issues.apache.org/jira/browse/IGNITE-14303

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/IndexValueCursor.java:
```diff
@@ -0,0 +1,67 @@
+
+        /** {@inheritDoc} */
+        @Override public Object get() {
+            return null;
```
It's implemented the way it was before. Please check GridH2Cursor.EMPTY [Link](https://github.com/apache/ignite/blob/2d2044a63440e9b9d8495f35819ccee26d8324e4/modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Cursor.java#L31)

timoninmaxim replied:
Unfortunately I can't fix this. H2Cursor for queries like "select MAX(col) from table" relies on this null. The same issue with SingleCursor. I can add comment there, and create a ticket to have this in mind for Calcite integration. WDYT?

timoninmaxim replied:
Create a ticket for that https://issues.apache.org/jira/browse/IGNITE-14303

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/sorted/IndexKey.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Represents a complex index key.
+ */
+public interface IndexKey {
```
Yes, I agree, this is wrong. As IndexKey expected to be used by user queries and it definitely should not contain info about CacheDataRow.

I've reworked a hierarchy of classes related to IndexKey, IndexRow. Now InlineIndexTree is generic of IndexRow, and not of IndexSearchRow (it was confusing, IMHO). IndexRow now extends IndexKey. As a result, I simplified a couple of places in the code.

timoninmaxim replied:
Hierarchy was simplified.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/SortedIndexSchema.java:
```diff
@@ -15,25 +15,40 @@
-    public FixedStringInlineIndexColumn(Column col, boolean useOptimizedCompare) {
-        super(col, Value.STRING_FIXED, useOptimizedCompare, false);
-    }
+    public IndexKeyDefinition[] getKeyDefinitions();
```
Schema and Definition are for different purposes. 

- IndexDefinition contains instructions on how to create an index and which parameters it has.
- Schema is responsible for mapping cache entry on index rows. 

So schema isn't just a list of keys. For example, `QueryIndexSchema` relies on the `GridH2RowDescriptor` object to provide the mapping. IOs and IndexRow require this mapping, but they actually should not have access to IndexDefinition.

Maybe just "Schema" is not a correct name for this target?

korlov42 replied:
> IndexDefinition contains instructions on how to create an index and which parameters it has

and with this in mind IndexDefinition is the best place for IndexKeyDefinition[]. Furthermore this is the way it's defined in IEP.

BTW why have you decided not to follow abstractions defined in IEP?

> Schema is responsible for mapping cache entry on index rows.

It's better to introduce something like this: 
```
public interface RowHandler<Row> {
    /** */
    Object get(int field, Row row);

    /** */
    int columnCount(Row row);
}
```

timoninmaxim replied:
Index keys moved to definition.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/cache/query/index/sorted/SortedIndex.java:
```diff
@@ -0,0 +1,94 @@
+     *
+     * @return amount of index tree segments.
+     */
+    public int segmentsCount();
```
> actually segments has nothing common with a generic sorted index

The `find` method already contains the segment argument in the signature. So `find` method of such general index will never be in use. For me when Ignite will need such general index it just extends existing one and hardcode segment with 0 value. WDYT?

korlov42 replied:
> it just extends existing one and hardcode segment with 0 value

Extension should brings new features, not limits them. And current approach violates the Interface Segregation Principle, I think

timoninmaxim replied:
I think, that it's not worth to add complexity to this hierarchy, as currently there is no non-segmented indexes. I've just renamed SortedIndex to SortedSegmentedIndex.

korlov42 reviewed (commented):
I have some doubts about refactoring of inlining columns. This introduces a few compatibility issues (dropped ability to treat POJO as a bytes array, as well as dropping supporting of fixed-size and case insensitive strings; with introduced support for modern Date/Time classes it's unclear to me how to differentiate between them and old ones for already created indexes) as well as closes doors for future improvement like [this one](https://issues.apache.org/jira/browse/IGNITE-13364).

timoninmaxim commented:
Hi @korlov42 ! Thanks a lot for reviewing my PR! You ask right questions. I worked on this PR pretty much time and then need to restore why I did some things this way. So I'm going to answer for one thing at a time.

> dropping supporting of fixed-size and case insensitive strings

Those 2 types actually aren't used in Ignite. 
1. There is a ticket for supporting case insensitive things [IGNITE-3999](https://issues.apache.org/jira/browse/IGNITE-3999). Also some comments in this ticket suggest to implement it in different way, with introducing functional indexes.
2. For fixed size strings. We work with them incorrectly and it looks like a bug. H2 provide a mapping: ValueStringFixed is used for char(), nchar(), character() types. But Ignite maps char() to String.class, then with H2Utils maps String.class to varchar, see [H2Utils.dbTypeFromClass](https://github.com/apache/ignite/blob/86073947248f0ca878e754e8b1b6181fdac72bd0/modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2Utils.java#L709)

I didn't have a problem to implement them, there will be just 2 classes that extends StringInlineIndexKeyType. But we can't test them now, as they will be not in use.

> closes doors for future improvement like [this one](https://issues.apache.org/jira/browse/IGNITE-13364).

For task IGNITE-13364, I check a PR (PR/8161)[https://github.com/apache/ignite/pull/8161/files] for this task and found that there is no problems to implement in with current changes. Actually there is `precision` field (GridH2RowDescriptor -> GridQueryProperty), that can be declared within IndexKeyDefinition, and used for computing inline size instead of parsing sql query.

So, why do you think it will close door for new improvements?

AMashenkov commented on modules/core/src/main/java/org/apache/ignite/internal/processors/cache/IgniteCacheOffheapManagerImpl.java:
```diff
@@ -2686,7 +2685,7 @@ private void finishUpdate(GridCacheContext cctx, CacheDataRow newRow, @Nullable
             GridCacheQueryManager qryMgr = cctx.queries();
 
-            if (qryMgr.enabled())
+            if (qryMgr.enabled() || cctx.kernalContext().indexing().enabled())
```
Is it possible to have a single flag for this?
Seems, it make no sense having Indexing enabled, but qryMgr disabled.

timoninmaxim replied:
> no sense having Indexing enabled, but qryMgr disabled.

Hi. It was a workaround, because for some reasons this condition was required at beginning of development. Currently I don't see in code any reasons to have this condition. So I'll remove it and check that everything is fine.

timoninmaxim replied:
Removed.

AMashenkov commented on modules/platforms/cpp/odbc-test/CMakeLists.txt:
```diff
@@ -62,7 +62,7 @@ set(SOURCES src/teamcity/teamcity_boost.cpp
         src/odbc_test_suite.cpp
         src/types_test.cpp
-        src/transaction_test.cpp
+#        src/transaction_test.cpp
```
Whats wrong with CPP tests?
Is there any ticket for fixing them?

timoninmaxim replied:
transaction_test.cpp uses TRANSACTIONAL_SNAPSHOT mode. As MVCC is deprecated I want to ignore this tests, but I've failed to find the right way to ignore them. So I comment them for a while. There are a lot of java tests that marked with Ignore now, I think can we just remove those tests at all? If yes I'd like to completely remove this CPP test too.

timoninmaxim commented on modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/IgniteH2Indexing.java:
```diff
@@ -464,41 +427,59 @@ public GridKernalContext kernalContext() {
+            tbl, wrappedCols.toArray(new IndexColumn[0]));
+
+        if (cacheInfo.affinityNode()) {
+            SortedIndexDefinition idxDef = new QueryIndexDefinition(
```
Hi @korlov42 ! Sorry for the time delay. I'm working on your comments. Most of the things look fine but there are some failed tests, I'm trying to fix them now. I hope, will push updated for all your comments today or on Monday.

I've lost compatibility for POJOs, my bad! I've added tests to the ignite-compatibility module and reproduced it, fix it. About other types (fixed strings, case insensitive strings) I wrote a comment below the PR, please check it.

I didn't implement it as `idxDefProvider`. As IndexDefinition should be instantiated before index creation, as IndexFactory accepts it as a parameter. So, I've moved the logic of finalizing key types to RowHandler. It replaced Schema as you suggest in the previous comment.

timoninmaxim replied:
Wait for merge of [#8811](https://github.com/apache/ignite/pull/8811).

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexTree.java:
```diff
@@ -0,0 +1,480 @@
+                    return 0;
+
+                // Other keys are not inlined. Should compare as rows.
+                if (i >= schema.getKeyDefinitions().length) {
```
Yes, the yardstick showed some degradation, but it wan't dramatically. I'd like to finalize that design is OK, then will fix issues with performance.

> worth to cache a slice of array of key definitions

Yes, I've implemented it. I will push the update as tests succeed.

timoninmaxim replied:
Replaced array with unmodified collection.

korlov42 reviewed (commented):
> Those 2 types actually aren't used in Ignite.

looks, you're right

> Actually there is `precision` field

approach with `precision` seems even better to me

> So, why do you think it will close door for new improvements?

I must admit I overreacted. I thought it's better to create a new instance for every column to have ability to store a column-specific information (like constraints, for instance), but now I see that we could derive the same info from index definition, and it's actually a better way.


And to the bad news: looks like inlining of POJO is completely broken. Please run `BasicJavaTypesIndexTest#testJavaPojoIndex`. The test actually passes, but we are not invoking `ObjectHashInlineIndexKeyType#compare0`. This could be fixed by removing [this](https://github.com/apache/ignite/pull/8490/files#diff-214234a606162716ee0b1a2045e17128517feadbc9819234a6d2af67c3021d2dR197) check, but I don't sure it won't bring another issues.

Also could you please describe relation between `IndexKey`, `JavaObjectKey`, and `NullKey`? A description of the `JavaObjectKey` states that all values could be wrapped, even an Integer. And this will work in case you will decide to check if both values an integer and a wrapped integer are equal. But there are several checks in the code that will not pass for wrapped values. Another problem is wrapping broke a comparability.

I think we need a better type hierarchy for key's values, something like H2's `Value`. It will help us avoid problems in the future.

korlov42 commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/DefaultIndexRowComparator.java:
```diff
@@ -0,0 +1,72 @@
+ * Default row comparator. Consider that every index key extends Comparable interface.
+ * Does not support comparation of different key types.
+ */
+public class DefaultIndexRowComparator implements IndexRowComparator {
```
is it used anywhere?

timoninmaxim replied:
Removed

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/IndexKeyImpl.java:
```diff
@@ -15,41 +15,48 @@
-     */
-    H2MvccExtrasInnerIO(short type, int ver, int payloadSize) {
-        super(type, ver, 28, payloadSize);
+public class IndexKeyImpl implements IndexKey {
```
Removed

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineRecommender.java:
```diff
@@ -0,0 +1,145 @@
+        boolean throttle = invokeCnt % inlineSizeThrottleThreshold != 0;
+
+        if (throttle)
+            return;
```
Hi, agree. Replaced with optimistic CAS.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/io/ThreadLocalSchemaHolder.java:
```diff
@@ -0,0 +1,43 @@
+ */
+public class ThreadLocalSchemaHolder {
+    /** */
+    private static final ThreadLocal<SortedIndexSchema> holder = new ThreadLocal<>();
```
Agree. Setup thread local with holder.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/keys/StringInlineIndexKeyType.java:
```diff
@@ -15,62 +15,42 @@
-    private final boolean useOptimizedCompare;
+    // TODO: what is it?
+//    /** Whether to use optimized comparison or not. */
+//    private final boolean useOptimizedCompare;
```
Restored previous behavior.

timoninmaxim commented on modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/index/QueryIndexSchema.java:
```diff
@@ -0,0 +1,210 @@
+                // columnValue ignores default columns (_KEY, _VAL), so make this shift.
+                return cacheDesc.columnValue(row.key(), row.value(), cacheIdx - QueryUtils.DEFAULT_COLUMNS_COUNT);
+        }
+    }
```
Agree. FIxed.

timoninmaxim commented:
Hi @korlov42 , @AMashenkov !

Thanks for all comments. I've fixed most of them, the biggest changes are:
1. Rework of hierarchies: IndexRow, IndexKey;
2. Hide packages from public to internal;
3. Restore backward compatibility for POJO type;
4. Replace Schema with RowHandler.
5. Introduce interfaces for date/time keys and delegates implementations to H2 (as it contains a lot of custom H2 logic, and license of h2 does not allow use this code, only in binary form).

For most of other comments I provide fixes, comment them and resolve issues. Some of issues are not resolved to just clearify that my fix is correct.

The only things to finalize:
1. Cursor implementations can't be implemented other way as H2 ("select max(col) from table" query) depends on the logic that it invokes get() before next(). Currently I see a single way - put a comment there and fix with Calcite integration. Or 
2. MVCC tests, including C++ mvcc test, can I just remove all of them?
3. If design is OK, then I will run performance tests.

Could you please review my PR again?

korlov42 commented on modules/core/src/main/java/org/apache/ignite/spi/indexing/IndexingSpi.java:
```diff
@@ -84,7 +91,56 @@
+     * @param newRow cache row to store in index.
+     * @param prevRow optional cache row that will be replaced with new row.
+     */
+    public void store(GridCacheContext<?, ?> cctx, CacheDataRow newRow, @Nullable CacheDataRow prevRow,
```
I don't sure it's a good idea to request an internal classes from public API

timoninmaxim replied:
Replace IgniteIndexing (IndexSpi) with IndexProcessor

korlov42 commented on modules/core/src/test/java/org/apache/ignite/internal/cache/query/index/sorted/inline/InlineIndexKeyTypeRegistryTest.java:
```diff
@@ -0,0 +1,88 @@
+    @Test
+    public void testNulls() {
+        int t = InlineIndexKeyTypeRegistry.get(NullIndexKey.INSTANCE, IndexKeyTypes.INT, pojoArrayKeyTypeSettings).type();
+        assert t == IndexKeyTypes.INT;
```
let's replace `assert` with JUnit assertions

korlov42 commented on modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/H2RowCache.java:
```diff
@@ -1,163 +0,0 @@
-/**
- * H2 row cache.
- */
-public class H2RowCache implements GridQueryRowCacheCleaner {
```
Why did you decide to get rid of this? This row cache is not the same thing as H2's internal object cache

timoninmaxim replied:
This was discussed with @tledkov-gridgain. It was an optimization for some specific corner cases for a specific GG client. Decided to remove this code as obsolete.

korlov42 replied:
I don't sure it's a good idea to remove row cache under within this PR. If the row cache is really useless, it's better to start a discussion on the dev list first.

korlov42 commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IgniteIndexing.java:
```diff
@@ -0,0 +1,329 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public void removeIndex(GridCacheContext<?, ?> cctx, IndexDefinition def, boolean softDelete) {
```
`IndexName` should uniquely describe index, so let's use it instead of `IndexDefinition`

korlov42 commented:
Hi, @timoninmaxim!

> Cursor implementations can't be implemented other way 

I'm OK with this

> MVCC tests, including C++ mvcc test, can I just remove all of them?

I think it's better to discuss with community

> If design is OK, then I will run performance tests.

Some refactoring still needed -- `IndexingSpi` should not accept classes from internal packages as methods params.

Also a few test classes were removed besides MVCC tests, but no replacement were provided. It's better to restore them.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/IndexSearchRow.java:
```diff
@@ -0,0 +1,36 @@
+/**
+ * Represents a complex index key.
+ */
+public interface IndexSearchRow {
```
Agree, for internal usage it is OK to merge them. The method `getKeys()` is used for defragmentation to skip extract values from an ignite cache twice.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/DurableBackgroundCleanupIndexTreeTask.java:
```diff
@@ -46,53 +49,44 @@
 
     /** */
-    private String cacheName;
+    private final IndexName idxName;
```
Hi, I will just remove Serializable. The reason to use it is a `DurableBackgroundCleanupIndexTreeTask`. It requires an index name for work. It was String, I replaced it with IndexName, then I will just revert this change.

timoninmaxim commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/keys/BooleanIndexKey.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    /** {@inheritDoc} */
+    @Override public int compare(IndexKey o, IndexKeyTypeSettings keySettings) {
```
Provided a SignedBytesIndexKey

timoninmaxim commented:
Hi @korlov42 ! Thanks for the review. I've fixed all your comments. Some details:
1. IndexName now is not Serializable. So I think there won't be any compatibility issues.
2. IgniteIndexing now is IndexProcessor. I reverted all changes in IndexSpi.
3. I added SignedBytesIndexKey to simplify the signature of the IndexKey.compare method.
4. I reviewed all deleted or ignored tests and found 2 that was ignored incorrectly (they test clean index tree task). Restored them.
5. The only entrypoint for index rows comparison is the `InlineIndexTree.compare` method. So I think it's OK that parts of the comparison mechanism are splitted between different classes, as all of them are used in single place.
6. Performance test is OK, no degradation for sql queries.

korlov42 commented on modules/indexing/src/test/java/org/apache/ignite/internal/processors/query/h2/database/inlinecolumn/InlineIndexColumnTest.java:
```diff
@@ -1,1014 +0,0 @@
- */
-@WithSystemProperty(key = "h2.serializeJavaObject", value = "false")
-@WithSystemProperty(key = "h2.objectCache", value = "false")
-public class InlineIndexColumnTest extends AbstractIndexingCommonTest {
```
Do we have replacement for this test?

timoninmaxim commented:
Hi @korlov42 ! I've updated the PR:
1. H2RowCache is restored as IndexRowCache;
2. InlineIndexColumnTest is restored. It helps me to find a bug, bad to miss this test from my side :(

Could you please check my PR again?

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/SortedSegmentedIndex.java:
```diff
@@ -0,0 +1,102 @@
+ */
+public interface SortedSegmentedIndex extends Index {
+    /**
+     * Finds index rows by specified range in specifed tree segment. Range can be bound or unbound.
```
typo: 'specifed' -> 'specified' and the same typo below

timoninmaxim replied:
fixed

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexFactory.java:
```diff
@@ -0,0 +1,34 @@
+ */
+public interface IndexFactory {
+    /**
+     * Creates index by specifed defition for specified cache.
```
typos

timoninmaxim replied:
fixed

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/SingleCursor.java:
```diff
@@ -0,0 +1,58 @@
+    /** Value to return */
+    private final T val;
+
+    /** Counter ot check wherther value is already got. */
```
typo 'wherther'

timoninmaxim replied:
fixed

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/SingleCursor.java:
```diff
@@ -0,0 +1,58 @@
+    }
+
+    /**
+     * Note that this implimentation violates the contract of GridCusror. It must be "before first" but this implementation
```
typo: 'implimentation'

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/defragmentation/DefragIndexFactory.java:
```diff
@@ -0,0 +1,262 @@
+        private final InlineIndexRowHandler rowHnd;
+
+        /** */
+        public BPlusInnerIoDelegate(IO io, InlineIndexRowHandler rowHnd) {
```
public constructor for private inner class (the same below)

timoninmaxim replied:
fixed.

timoninmaxim replied:
fixed.

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/io/AbstractLeafIO.java:
```diff
@@ -0,0 +1,85 @@
+    AbstractLeafIO(int type, int ver, int itemSize) {
+        super(type, ver, itemSize);
+    }
+
```
empty line

timoninmaxim replied:
fixed.

tledkov-gridgain commented on modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/sorted/inline/types/DateValueUtils.java:
```diff
@@ -0,0 +1,43 @@
+ *
+ * dv = (year << SHIFT_YEAR) | (month << SHIFT_MONTH) | day.
+ */
+public class DateValueUtils {
```
Should we use interface to hold constants or add private constructor?

timoninmaxim replied:
Add private constructor. Rename DateValueUtils to DateValueConstants.
